modules/core/src/main/java/org/apache/ignite/internal/causality/VersionedValue.java

```diff
@@ -0,0 +1,220 @@
+     *                                  every update of storage revision as a listener.
+     */
+    public VersionedValue(
+            @Nullable BiConsumer<VersionedValue<T>, Long> onStorageRevisionUpdate,
+            Consumer<Consumer<Long>> observableRevisionUpdater
+    ) {
+        this(null, observableRevisionUpdater, 2);
+    }
+
+    /**
+     * Constructor with default history size that equals 2 and no closure.
+     *
+     * @param observableRevisionUpdater A closure intended to connect this VersionedValue with a revision updater, that this VersionedValue
+     *                                  should be able to listen to, for receiving storage revision updates. This closure is called once on
+     *                                  a construction of this VersionedValue and accepts a {@link Consumer<Long>} that should be called on
+     *                                  every update of storage revision as a listener.
+     */
+    public VersionedValue(Consumer<Consumer<Long>> observableRevisionUpdater) {
+        this(null, observableRevisionUpdater);
+    }
+
+    /**
+     * Creates a future for this value and causality token, or returns it if it already exists.
+     *
+     * <p>The returned future is associated with an update having the given causality token and completes when this update is finished
+     * applying.
+     *
+     * @param causalityToken Causality token. Let's assume that the update associated with token N is already applied to this value. Then,
+     *                       if token N is given as an argument, a completed future will be returned. If token N - 1 is given, this method
+     *                       returns the result in the state that is actual for the given token. If the token is strongly outdated, {@link
+     *                       OutdatedTokenException} is thrown. If token N + 1 is given, this method will return a future that will be
+     *                       completed when the update associated with token N + 1 will have been applied. Tokens that greater than N by
+     *                       more than 1 should never be passed.
+     * @return The future.
+     * @throws OutdatedTokenException If outdated token is passed as an argument.
+     */
+    public CompletableFuture<T> get(long causalityToken) throws OutdatedTokenException {
+        long actualToken0 = actualToken;
+
+        if (causalityToken <= actualToken0) {
+            Entry<Long, CompletableFuture<T>> histEntry = history.floorEntry(causalityToken);
+
+            if (histEntry == null) {
+                throw new OutdatedTokenException(causalityToken, actualToken0, historySize);
+            }
+
+            return histEntry.getValue();
+        } else {
+            var fut = new CompletableFuture<T>();
+
+            CompletableFuture<T> previousFut = history.putIfAbsent(causalityToken, fut);
```

Comment:
What if following happens:
1. Thread1 enters #get, finds out that the given causality token is greater than actualToken and enters `else` section
2. Thread2 makes multiple revision updates and actualToken becomes greater than causality token given for Thread1, it also trims history
3. Thread1 puts new future to the history with an obsolete token

Seems that Thread1 will receive a future that will never complete. This can lead to rare but misterious bugs

Reply:
Fixed with RWLock. The lock do not allow invoking a trim history during getting a value.